pkg/logproto/compat.go

```diff
@@ -214,3 +215,65 @@ func init() {
 	jsoniter.RegisterTypeEncoderFunc("logproto.LegacySample", SampleJsoniterEncode, func(unsafe.Pointer) bool { return false })
 	jsoniter.RegisterTypeDecoderFunc("logproto.LegacySample", SampleJsoniterDecode)
 }
+
+func MergeLabelResponses(responses []*LabelResponse) (*LabelResponse, error) {
+	if len(responses) == 0 {
+		return nil, fmt.Errorf("no label responses to merge")
+	} else if len(responses) == 1 {
+		return responses[0], nil
+	}
+
+	result := &LabelResponse{
+		Values: make([]string, 0, len(responses)),
+	}
+
+	unique := make(map[string]bool)
+
+	for _, r := range responses {
+		for _, v := range r.Values {
+			if _, ok := unique[v]; !ok {
+				unique[v] = true
+				result.Values = append(result.Values, r.Values...)
+			} else {
+				continue
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// TODO(jordanrushing): Fix this / improve performance (it's so slow); is there a better way to return only the combination of unique SeriesIdentifier?
+func MergeSeriesResponses(responses []*SeriesResponse) (*SeriesResponse, error) {
+	if len(responses) == 0 {
+		return nil, fmt.Errorf("no series responses to merge")
+	} else if len(responses) == 1 {
+		return responses[0], nil
+	}
+
+	// Hash the response struct and compare to avoid duplicates
+	hashString := func(r *SeriesResponse) string {
+		h := sha256.New()
+		h.Write([]byte(fmt.Sprintf("%v", r)))
+		return fmt.Sprintf("%x", h.Sum(nil))
+	}
```

Comment:
I assume this is because you use a cryptographic hash which is often slower. In any case. Check out the de-duplication in the querier (https://github.com/grafana/loki/blob/f6d15fba48c5c69bd28ede308810f77db057a6e0/pkg/querier/querier.go#L435-L443). You could factor this out and we can benchmark it later if required.